### PR TITLE
9 oauth2 authentication using multiple providers

### PR DESCRIPTION
**Pull request by Tomislaw**
Added Oauth2 authentication using 3 providers: Discord, Google and Github.
There are some further tests needed but there shoudn't be any major changes in the code.

### CODE REVIEW

File: modules/odr_api/odr_api/api/endpoints/auth.py [hunk 1]
```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
...
+    existing_user = user_crud.get_user_by_email(db, openid_user.email)
+    if not existing_user:
+        existing_user = user_crud.create_user_from_openid(db, openid_user)
+
+    session = user_crud.login_openid_user(db, openid_user)
+    if not session:
+        raise HTTPException(status_code=401, detail="Different provider already used")
+
+    response = RedirectResponse(url=f"{request.base_url}{settings.OAUTH2_REDIRECT_PATH}")
+    response.set_cookie(key="session", value=session.id, httponly=True)
```
Comment (CheesyLaZanya) [inline modules/odr_api/odr_api/api/endpoints/auth.py:94]: Should this have secure=True as well? Or what are the expectations there?

File: modules/odr_api/odr_api/api/endpoints/auth.py [hunk 2]
```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
...
+
+    return response
+
+
+# Google SSO
+google_sso = GoogleSSO(
+    client_id=settings.GOOGLE_CLIENT_ID,
+    client_secret=settings.GOOGLE_CLIENT_SECRET,
+    redirect_uri="http://localhost:31100/api/v1/auth/google/callback",
```
Comment (CheesyLaZanya) [inline modules/odr_api/odr_api/api/endpoints/auth.py:103]: This is hardcoded as localhost, should it be?

File: modules/odr_api/odr_api/api/endpoints/auth.py [hunk 3]
```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
...
 
 @router.get("/auth/logout/all")
-def logout_all(db=Depends(get_db), current_user: User = Depends(AuthProvider())):
+def logout_all(current_user: User = Depends(AuthProvider()), db: Session = Depends(get_db)):
     user_crud.delete_user_sessions(db, current_user.id)
 
     return UserLogout()
+
+
+def ouath2_login_and_signup(
```
Comment (CheesyLaZanya) [inline modules/odr_api/odr_api/api/endpoints/auth.py:77]: Typo in "oauth2"

File: modules/odr_api/odr_api/api/endpoints/auth.py [hunk 4]
```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
...
+        except Exception:
+            raise HTTPException(status_code=401, detail="Failed to login with Google")
+        return ouath2_login_and_signup(request, user, db)
+
+
+# Github SSO
+github_sso = GithubSSO(
+    client_id=settings.GITHUB_CLIENT_ID,
+    client_secret=settings.GITHUB_CLIENT_SECRET,
+)
```
Comment (CheesyLaZanya) [inline modules/odr_api/odr_api/api/endpoints/auth.py:129]: Does Github SSO not need a redirect uri to api/v1/auth/github/callback similar to Google?

File: modules/odr_api/odr_api/api/endpoints/auth.py [hunk 5]
```diff
@@ -46,7 +68,139 @@ def logout(db=Depends(get_db), session_cookie: str = Depends(get_session_cookie)
...
+    default_scope=["identify", "email"],
+    discovery_document=discovery_document,
+    response_convertor=convert_discord_openid,
+)
+
+
+discord_sso = DiscordSSO(
+    client_id=settings.DISCORD_CLIENT_ID,
+    client_secret=settings.DISCORD_CLIENT_SECRET,
+)
```
Comment (CheesyLaZanya) [inline modules/odr_api/odr_api/api/endpoints/auth.py:188]: Does Discord SSO not need a redirect uri to api/v1/auth/discord/callback similar to Google?

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 2 above)
Comment (fearnworks) [reply modules/odr_api/odr_api/api/endpoints/auth.py:103]: Should not be. Will break inside of the containers.

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 4 above)
Comment (Tomislaw) [reply modules/odr_api/odr_api/api/endpoints/auth.py:129]: The github configuration was the wrong one.
Redirect uri is calculated automatically using this method. So this property is not needed.
```
        return await github_sso.get_login_redirect(
            redirect_uri=request.url_for("github_callback")
        )
```

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 5 above)
Comment (Tomislaw) [reply modules/odr_api/odr_api/api/endpoints/auth.py:188]: The github configuration was the wrong one.
Redirect uri is calculated automatically using this method. So this property is not needed.
```
        return await discord_sso.get_login_redirect(
            redirect_uri=request.url_for("discord_callback")
        )
```

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 1 above)
Comment (Tomislaw) [reply modules/odr_api/odr_api/api/endpoints/auth.py]: Fixed.

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 2 above)
Comment (Tomislaw) [reply modules/odr_api/odr_api/api/endpoints/auth.py]: Removed this property as it was not needed. Redirect uri is already set in different place.

File: modules/odr_api/odr_api/api/endpoints/auth.py (hunk 3 above)
Comment (Tomislaw) [reply modules/odr_api/odr_api/api/endpoints/auth.py]: Fixed.

**Review by CheesyLaZanya — APPROVED** [review]
Updates for the issues I pointed out look good.

**Pull request merged**